ssl/statem/extensions.c

```diff
@@ -0,0 +1,974 @@
+                           PACKET_remaining(&currext->data),
+                           s->tlsext_debug_arg);
+
+    /* Skip if we've already parsed this extension */
+    if (currext->parsed)
+        return 1;
+
+    currext->parsed = 1;
+
+    if (idx < OSSL_NELEM(ext_defs)) {
+        /* We are handling a built-in extension */
+        const EXTENSION_DEFINITION *extdef = &ext_defs[idx];
+
+        /* Check if extension is defined for our protocol. If not, skip */
+        if (!extension_is_relevant(s, extdef->context, context))
+            return 1;
+
+        parser = s->server ? extdef->parse_client_ext : extdef->parse_server_ext;
+
+        if (parser != NULL) {
+            if (!parser(s, &currext->data, al))
+                return 0;
+
+            return 1;
+        }
+
+        /*
+         * If the parser is NULL we fall through to the custom extension
+         * processing
+         */
+    }
+
+    /*
+     * This is a custom extension. We only allow this if it is a non
+     * resumed session on the server side.
+     *
+     * TODO(TLS1.3): We only allow old style <=TLS1.2 custom extensions.
+     * We're going to need a new mechanism for TLS1.3 to specify which
+     * messages to add the custom extensions to.
+     */
+    if ((!s->hit || !s->server)
+            && (context
+                & (EXT_CLIENT_HELLO | EXT_TLS1_2_SERVER_HELLO)) != 0
+            && custom_ext_parse(s, s->server, currext->type,
+                                PACKET_data(&currext->data),
+                                PACKET_remaining(&currext->data),
+                                al) <= 0)
+        return 0;
+
+    return 1;
+}
+
+/*
+ * Parse all remaining extensions that have not yet been parsed. Also calls the
+ * finalisation for all extensions at the end, whether we collected them or not.
+ * Returns 1 for success or 0 for failure. On failure, |*al| is populated with a
+ * suitable alert code.
+ */
+int tls_parse_all_extensions(SSL *s, int context, RAW_EXTENSION *exts, int *al)
+{
+    size_t loop, numexts = OSSL_NELEM(ext_defs);
+
+    /* Calculate the number of extensions in the extensions list */
+    if ((context & EXT_CLIENT_HELLO) != 0) {
+        numexts += s->cert->srv_ext.meths_count;
+    } else if ((context & EXT_TLS1_2_SERVER_HELLO) != 0) {
+        numexts += s->cert->cli_ext.meths_count;
+    }
+
+    /* Parse each extension in turn */
+    for (loop = 0; loop < numexts; loop++) {
+        if (!tls_parse_extension(s, loop, context, exts, al))
+            return 0;
+    }
+
+    /*
+     * Finalise all known extensions relevant to this context, whether we have
+     * found them or not
+     */
+    for (loop = 0; loop < OSSL_NELEM(ext_defs); loop++) {
+        if(ext_defs[loop].finalise_ext != NULL
+                && (ext_defs[loop].context & context) != 0
+                && !ext_defs[loop].finalise_ext(s, context, exts[loop].present,
+                                                al))
+            return 0;
```

Comment:
I disagree. Its significant additional complexity...what happens when you get an error half way through running the initialisers (run the finalisers but only for the ones we initialised so far?); what happens if you get an error from a finaliser if we already got an error earlier? That kind of complexity is likely to be difficult to get right...I really don't want to implement it when I know we don't even need it right now. However, on the other hand, I can foresee that at some point in the future when we update the custom extension API to work for TLS1.3, we might need to do something like that for that new API. I want to cross that bridge when we come to it (when I know we really do need to solve it, rather than guessing that we might).